ostd/src/mm/page/meta.rs

```diff
@@ -306,18 +311,15 @@ pub(crate) fn init() -> Vec<Page<MetaPageMeta>> {
 
 fn alloc_meta_pages(nframes: usize) -> Vec<Paddr> {
     let mut meta_pages = Vec::new();
-    let start_frame = allocator::PAGE_ALLOCATOR
-        .get()
-        .unwrap()
-        .lock()
-        .alloc(core::alloc::Layout::from_size_align(nframes * PAGE_SIZE, PAGE_SIZE).unwrap())
-        .expect("Failed to allocate metadata pages");
-    // Zero them out as initialization.
-    let vaddr = paddr_to_vaddr(start_frame) as *mut u8;
-    unsafe { core::ptr::write_bytes(vaddr, 0, PAGE_SIZE * nframes) };
-    for i in 0..nframes {
-        let paddr = start_frame + i * PAGE_SIZE;
-        meta_pages.push(paddr);
+    let mut allocator = BOOTSTRAP_PAGE_ALLOCATOR.get().unwrap().lock();
```

Comment:
Here, we are using `BOOTSTRAP_PAGE_ALLOCATOR` explicitly.

```rust
    let mut meta_pages = Vec::new();
```

But in the first line of this function, we are doing heap allocation by using `Vec`. Where does the pages used by heap come from? Do we have any mechanism that prevents the heap allocator from using `PAGE_ALLOCATOR`?

Reply:
According to the previous implementation, the initial space managed by the heap allocator comes from a const vector sized 256 pages:

``` rust
const INIT_KERNEL_HEAP_SIZE: usize = PAGE_SIZE * 256;

#[repr(align(4096))]
struct InitHeapSpace([u8; INIT_KERNEL_HEAP_SIZE]);

static mut HEAP_SPACE: InitHeapSpace = InitHeapSpace([0; INIT_KERNEL_HEAP_SIZE]);
```

The heap allocator will function correctly before the initial space is burnt out. If all the initial space is used, the heap allocator will call its `rescue` function to apply new pages from the page allocator. 

The kernel may crash if `PAGE_ALLOCATOR` is not ready while the heap needs new pages. Since only `meta` will use heap memory before `PAGE_ALLOCTOR` is initialized, while the memory usage of meta can be estimated, it is of small possibility that the kernel will crash due to the aforementioned reasons.